core/src/main/java/com/amplitude/core/events/Identify.kt

```diff
@@ -15,10 +15,21 @@ enum class IdentifyOperation(val operationType: String) {
     REMOVE("\$remove")
 }
 
-open class Identify() {
+open class Identify {
 
     private val propertySet: MutableSet<String> = mutableSetOf()
-    val properties = mutableMapOf<String, Any?>()
+
+    private val _properties = mutableMapOf<String, Any?>()
```

Comment:
I think this will fix the first error:
`com.amplitude.analytics.connector.IdentityStoreImpl$editIdentity$1.updateUserProperties(IdentityStore.kt:75)
com.amplitude.android.plugins.AnalyticsConnectorPlugin.execute(AnalyticsConnectorPlugin.kt:45)`

but I still think there still could be errors at:
`com.amplitude.android.plugins.AnalyticsConnectorPlugin.execute(AnalyticsConnectorPlugin.kt:35)`.

Maybe would it be enough to clone the userProperties in `AnalyticsConnectorPlugin.kt` line 30:
` val eventUserProperties: Map<String, Any?>? = event.userProperties?.toMap()`

Reply:
Do we need to make Identify object thread safe? Legacy Android didn't have any synchronization, would be nice to keep our base objects simple too.

I'm not sure if this is limited to Identity but maybe we can instead update the  `IdentityStoreImpl` in Analytics Connector. It already has synchronization code. wdyt?
https://github.com/amplitude/experiment-android-client/blob/main/analytics-connector/src/main/java/com/amplitude/analytics/connector/IdentityStore.kt#L37

Reply:
> but I believe there still could be errors at:
com.amplitude.android.plugins.AnalyticsConnectorPlugin.execute(AnalyticsConnectorPlugin.kt:35).

I was able to reproduce the error on `main` and now it is fixed.
I clone 2 levels (maybe ideally we should clone whole Identify object):
```
val clonedProperties = _properties.toMutableMap() // 1st level
for ((key, value) in clonedProperties) {
    if (value is Map<*, *>) {
        clonedProperties[key] = value.toMutableMap() // 2nd level
    }
}
return clonedProperties
```
1st level cloning fixes  `com.amplitude.android.plugins.AnalyticsConnectorPlugin.execute(AnalyticsConnectorPlugin.kt:35)`
2nd level cloning -  `com.amplitude.analytics.connector.IdentityStoreImpl$editIdentity$1.updateUserProperties(IdentityStore.kt:75)`

> Do we need to make Identify object thread safe? Legacy Android didn't have any synchronization, would be nice to keep these base objects simple too.

In this case we should mention in documentation that an Identify object is not thread-safe and should not be modified after sending to Amplitude.
If developers want to share an Identify object and modify it later they have to implement "cloning" manually.

> Maybe would it be enough to clone the userProperties in AnalyticsConnectorPlugin.kt line 30:
 val eventUserProperties: Map<String, Any?>? = event.userProperties?.toMap()
 I'm not sure if this is limited to Identity but maybe we can instead update the IdentityStoreImpl in Analytics Connector. It already has synchronization code. wdyt?
 
I think, the root issue is [line](https://github.com/amplitude/Amplitude-Kotlin/blob/main/core/src/main/java/com/amplitude/core/Amplitude.kt#L190)
```
event.userProperties = identify.properties
```
Here we set `event.userProperties` to be the same object as `identify.properties`. Then the event is sent asynchronously to another thread/coroutine and `identify.properties` subsequent modifications in main thread can raise the ConcurrentModificationException. For example, in `val eventUserProperties: Map<String, Any?>? = event.userProperties?.toMap()`  ConcurrentModificationException can be raised when the original map `event.userProperties (=== identify.properties)` is enumerated in `toMap()` in the same time when `identify.properties` is modified.

Reply:
I guess current way could help, though I wonder if leaving the identify as it is now and clone on `event.userProperties = identify.properties` could also work. Ideally the identify object wouldn't have any impact after they send to our method.

Reply:
@qingzhuozhen 
The `identify.properties` is read in a few locations:
```
fun identify(identify: Identify, options: EventOptions? = null): Amplitude {
    ...
    event.userProperties = identify.properties
    ...
}

fun groupIdentify(groupType: String, groupName: String, identify: Identify, options: EventOptions? = null): Amplitude {
    ...
    event.groupProperties = identify.properties
    ...
}

fun setGroup(groupType: String, groupName: String, options: EventOptions? = null): Amplitude {
    ...
    userProperties = identify.properties
    ...
}

fun setGroup(groupType: String, groupName: Array<String>, options: EventOptions? = null): Amplitude {
    ...
    userProperties = identify.properties
    ...
}
```

Reply:
Gotcha, then it might be easier to do it on get.